cranelift/wasm/src/module_translator.rs

```diff
@@ -18,72 +18,108 @@ pub fn translate_module<'data>(
 ) -> WasmResult<ModuleTranslationState> {
     let _tt = timing::wasm_translate_module();
     let mut module_translation_state = ModuleTranslationState::new();
+    let mut validator = Validator::new();
+    validator.wasm_features(environ.wasm_features());
 
     for payload in Parser::new(0).parse_all(data) {
         match payload? {
-            Payload::Version { .. } | Payload::End => {}
+            Payload::Version { num, range } => {
+                validator.version(num, &range)?;
+            }
+            Payload::End => {
+                validator.end()?;
+            }
 
             Payload::TypeSection(types) => {
+                validator.type_section(&types)?;
```

Comment:
(Probably a dumb question, but just a sanity check) Does the validator keep internal state to remember which sections it has seen? This might have an impact since sections are supposed to appear in a certain order.

Reply:
It does indeed!